src/services/codefixes/fixUnmatchedParameter.ts

```diff
@@ -0,0 +1,102 @@
+    const renameUnmatchedParameter = "renameUnmatchedParameter";
+
+    const errorCodes = [
+        Diagnostics.JSDoc_param_tag_has_name_0_but_there_is_no_parameter_with_that_name.code,
+    ];
+
+    registerCodeFix({
+        fixIds: [deleteUnmatchedParameter, renameUnmatchedParameter],
+        errorCodes,
+        getCodeActions: function getCodeActionsToFixUnmatchedParameter(context) {
+            const { sourceFile, span } = context;
+            const actions: CodeFixAction[] = [];
+            const info = getInfo(sourceFile, span.start);
+            if (info) {
+                append(actions, getDeleteAction(context, info));
+                append(actions, getRenameAction(context, info));
+                return actions;
+            }
+            return undefined;
+        },
+        getAllCodeActions: function getAllCodeActionsToFixUnmatchedParameter(context) {
+            const tagsToSignature = new Map<SignatureDeclaration, JSDocTag[]>();
+            return createCombinedCodeActions(textChanges.ChangeTracker.with(context, changes => {
+                eachDiagnostic(context, errorCodes, ({ file, start }) => {
+                    const info = getInfo(file, start);
+                    if (info) {
+                        tagsToSignature.set(info.signature, append(tagsToSignature.get(info.signature), info.jsDocParameterTag));
+                    }
+                });
+
+                tagsToSignature.forEach((tags, signature) => {
+                    if (context.fixId === deleteUnmatchedParameter) {
+                        const tagsSet = new Set(tags);
+                        changes.filterJSDocTags(signature.getSourceFile(), signature, t => !tagsSet.has(t));
+                    }
+                });
+            }));
+        }
+    });
+
+    function getDeleteAction(context: CodeFixContext, { name, signature, jsDocParameterTag }: Info) {
+        const changes = textChanges.ChangeTracker.with(context, changeTracker =>
+            changeTracker.filterJSDocTags(context.sourceFile, signature, t => t !== jsDocParameterTag));
+        return createCodeFixAction(
+            deleteUnmatchedParameter,
+            changes,
+            [Diagnostics.Delete_unused_param_tag_0, name.getText(context.sourceFile)],
+            deleteUnmatchedParameter,
+            Diagnostics.Delete_all_unused_param_tags
+        );
+    }
+
+    function getRenameAction(context: CodeFixContext, { name, signature, jsDocParameterTag }: Info) {
+        if (!length(signature.parameters)) return undefined;
+
+        const sourceFile = context.sourceFile;
+        const tags = getJSDocTags(signature);
+        const names = new Set<__String>();
+        for (const tag of tags) {
+            if (isJSDocParameterTag(tag) && isIdentifier(tag.name)) {
+                names.add(tag.name.escapedText);
+            }
+        }
+        const parameterName = firstDefined(signature.parameters, p =>

```

Comment:
if there are two unmatched `@param` and two unmatched parameters, both tags will initially offer to rewrite to the first parameter, right? For example:

```ts
/**
 * @param old
 * @param typod
 * @param ok
 * @param good
 */
function f(ok, good, alt, typo) { }
```

I think this is OK since fixing the first will cause the second to switch to the second unmatched parameter, but we should have a test for it, plus a comment pointing out that it could be improved later.

Reply:
@sandersn Oke. I've added tests.